examples/10.dynamic_configuration/config_broker.cc

```diff
@@ -0,0 +1,275 @@
+	static uint16_t nextId = 1;
+
+	ConfigToken *token =
+	  token_unseal<ConfigToken>(key, Sealed<ConfigToken>{sealed_cap});
+
+	if (token == nullptr)
+	{
+		Debug::log("invalid config capability {}", sealed_cap);
+		return nullptr;
+	}
+
+	if (token->id == 0)
+	{
+		// Assign an ID so we can track the callbacks added
+		// from this capability
+		token->id = nextId++;
+
+		// Count how many config ids are defined to
+		// save us having to do this each time
+		for (auto i = 0; i < MAX_CONFIG_IDS; i++)
+		{
+			if (strlen(token->configId[i]))
+			{
+				token->count++;
+			}
+		}
+	}
+
+	return token;
+}
+
+//
+// Find a Config by name.  If it doesn't already exist
+// create one.
+//
+Config *find_config(const char *name)
+{
+	for (auto &c : configData)
+	{
+		if (strcmp(c->name, name) == 0)
+		{
+			return c;
+		}
+	}
+
+	// Allocate a Config object
+	Config *c = static_cast<Config *>(malloc(sizeof(Config)));
+
+	// Save the name
+	c->name = static_cast<char *>(malloc(strlen(name)));
+	strncpy(c->name, name, strlen(name));
+
+	c->updated = false;
+	c->data    = nullptr;
+
+	// Add it to the vector
+	configData.push_back(c);
+
+	return c;
+};
+
+//
+// Add a callback to the list for a Config.  Use the
+// id to prevent adding more than one callback for
+// each client
+//
+void add_callback(Config               *c,
+                  uint16_t              id,
+                  __cheri_callback void cb(const char *, void *))
+{
+	for (auto &cb_info : c->cbList)
+	{
+		if (cb_info->id == id)
+		{
+			cb_info->cb = cb;
+			return;
+		}
+	}
+
+	cbInfo *cb_info = static_cast<cbInfo *>(malloc(sizeof(cbInfo)));
+	cb_info->id     = id;
+	cb_info->cb     = cb;
+	c->cbList.push_back(cb_info);
+}
+
+//
+// Set a configuration data item.  The capabailty must have the
+// is_source set to true and include the name of the item being
+// set
+//
+void __cheri_compartment("config_broker")
+  set_config(SObj sealed_cap, const char *name, void *data)
+{
+	ConfigToken *token = config_capability_unseal(sealed_cap);
+
+	if (!token->is_source)
+	{
+		Debug::log("Not a source capability: {}", sealed_cap);
+		return;
+	}
+
+	bool valid = false;
+	for (auto i = 0; i < token->count; i++)
+	{
+		if (strcmp(token->configId[i], name) == 0)
+		{
+			valid = true;
+			break;
+		}
+	}
+	if (!valid)
+	{
+		Debug::log("Not a source capability for: {} {}", sealed_cap, name);
+		return;
+	}
+
+	// See if we already have a config structure
+	Config *c = find_config(name);
+
+	// Free any claim we had on the previous data
+	if (c->data)
+	{
+		heap_free(MALLOC_CAPABILITY, c->data);
+	}
+
+	// Create a read only version of the capability to pass to
+	// the subscribers, and add a claim so we can deliver it
+	// when a compartment registers a callback.
+	CHERI::Capability readOnlyData{data};
+	readOnlyData.permissions() &=
+	  {CHERI::Permission::Load, CHERI::Permission::Global};
+	c->data = readOnlyData;
+	heap_claim(MALLOC_CAPABILITY, c->data);
```

Comment:
We should probably move this to a separate issue / discussion but:

I would like to make quota objects a bit more dynamic.  I want to be able to take my quota and create a new sub-quota that can be passed to another compartment allowing them to allocate a specific amount.  I'd also like it to be able to restrict to allocation, claims, or both, and maybe not allow deallocation (or only allow deallocation).

With a mechanism like that, we could do a little dance of:

 - Allocate a new n-byte quota from the current one.
 - Free the old object.
 - Claim with the subquota.
 - Destroy the subquota (returning unused space to the parent)

The main tricky part is that `heap_free_all` will not reclaim subquotas and so may need a mechanism for allowing only temporary / reclaimable subquotas.  I'd prefer to build a generic mechanism that can be used to build the atomically replace claim mechanism (because it would also be nice to have things like atomically allocate and claim with this other quota for some use cases and possibly even atomically allocate and claim with this set of quotas) than build that mechanism in the allocator, but I'm not sure if it's possible.